From d054e6f9e28d160811b19deb834912eb69f59376 Mon Sep 17 00:00:00 2001
From: Tanguy Pruvot <tanguy.pruvot@gmail.com>
Date: Fri, 27 Dec 2013 00:55:27 +0100
Subject: [PATCH] hwui: Allow to blacklist android applications

Add support for a local config folder with blacklisted process names
/data/local/hwui.deny/<package name>

Only enabled with "hwui.use.blacklist=true"

This system can be used easily from adb shell, without reboot
or with a setting app, like the one i use, HwaSettings

Change-Id: I436b5dd7d6bf9fd36be7e86fc9d2a174a66701ff
---
 core/java/android/app/ActivityThread.java | 33 +++++++++++++++++++++++++++++++
 1 file changed, 33 insertions(+)

diff --git a/core/java/android/app/ActivityThread.java b/core/java/android/app/ActivityThread.java
index cab7d83..2a68fe4 100644
--- a/core/java/android/app/ActivityThread.java
+++ b/core/java/android/app/ActivityThread.java
@@ -4126,6 +4126,37 @@ public final class ActivityThread {
         }
     }
 
+    /**
+     * hwui.use.blacklist allows to disable the hardware acceleration
+     * to specified applications processes, if files (process names)
+     * are present in /data/local/hwui.deny/
+     */
+    private boolean hwuiForbidden(String processName) {
+
+        boolean useBL = SystemProperties.getBoolean("hwui.use.blacklist", false);
+
+        // Default is allowed
+        boolean blacklisted = false;
+
+        if (!useBL || TextUtils.isEmpty(processName))
+            return blacklisted;
+
+        File hwuiConfig = new File("/data/local/hwui.deny/" + processName);
+        if (hwuiConfig.exists()) {
+            blacklisted = true;
+        }
+
+        hwuiConfig = null;
+
+        // Keep the logs to show process names with "adb logcat | grep listed"
+        if (!blacklisted)
+            Slog.v(TAG, processName + " white listed for hwui");
+        else
+            Slog.d(TAG, processName + " black listed for hwui");
+
+        return blacklisted;
+    }
+
     private void handleBindApplication(AppBindData data) {
         mBoundApplication = data;
         mConfiguration = new Configuration(data.config);
@@ -4178,6 +4209,8 @@ public final class ActivityThread {
             if (!ActivityManager.isHighEndGfx()) {
                 HardwareRenderer.disable(false);
             }
+        } else if (hwuiForbidden(data.processName)) {
+            HardwareRenderer.disable(false);
         }
         
         if (mProfiler.profileFd != null) {
-- 
1.8.3.2

